File: versao-java/src/test/java/api/produtos/java/IntegrationTests.java
package api.produtos.java;

import api.produtos.java.entity.Produto;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.ResponseEntity;
import org.springframework.test.annotation.DirtiesContext;

import java.math.BigDecimal;

import static api.produtos.java.assertions.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.OK;
import static org.springframework.http.RequestEntity.*;
import static org.springframework.test.annotation.DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD;

@SpringBootTest(webEnvironment = RANDOM_PORT)
@DirtiesContext(classMode = AFTER_EACH_TEST_METHOD)
public class IntegrationTests {
    @Autowired
    private TestRestTemplate restTemplate;

    @Test
    public void deveCriarNovoProduto() {
        Produto produto = new Produto(null, "Laptop", new BigDecimal("2400.00"));

        ResponseEntity<Produto> resposta = restTemplate.exchange(
                post("/produtos").body(produto), Produto.class);

        assertThat(resposta.getStatusCode()).isEqualTo(CREATED);
        assertThat(resposta.getBody())
                .hasId(3L)
                .hasNome("Laptop")
                .hasPreco(new BigDecimal("2400.00"));
    }

    @Test
    public void deveBuscarProdutos() {
        ResponseEntity<Produto[]> resposta = restTemplate.exchange(
                get("/produtos").build(), Produto[].class);

        assertThat(resposta.getStatusCode()).isEqualTo(OK);
        assertThat(resposta.getBody().length).isEqualTo(2);
        assertThat(resposta.getBody()[0])
                .hasId(1L)
                .hasNome("TV")
                .hasPreco(new BigDecimal("1600.00"));
        assertThat(resposta.getBody()[1])
                .hasId(2L)
                .hasNome("Display")
                .hasPreco(new BigDecimal("800.00"));
    }

    @Test
    public void deveAtualizarProduto() {
        Produto produto = new Produto(1L, "Laptop", new BigDecimal("2400.00"));

        ResponseEntity<Produto> resposta = restTemplate.exchange(
                put("/produtos/1").body(produto), Produto.class);

        assertThat(resposta.getStatusCode()).isEqualTo(OK);
        assertThat(resposta.getBody())
                .hasId(1L)
                .hasNome("Laptop")
                .hasPreco(new BigDecimal("2400.00"));
    }

    @Test
    public void deveDeletarProduto() {
        ResponseEntity<Produto> resposta = restTemplate.exchange(
                delete("/produtos/1").build(), Produto.class);

        assertThat(resposta.getStatusCode()).isEqualTo(OK);
        assertThat(resposta.getBody())
                .hasId(1L)
                .hasNome("TV")
                .hasPreco(new BigDecimal("1600.00"));
    }
}

File: versao-kotlin/src/test/kotlin/api/produtos/kotlin/IntegrationTests.kt
package api.produtos.kotlin

import api.produtos.kotlin.assertions.Assertions.assertThat
import api.produtos.kotlin.entity.Produto
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.test.web.client.getForEntity
import org.springframework.boot.test.web.client.postForEntity
import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.HttpStatus.OK
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.annotation.DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD
import java.math.BigDecimal

@SpringBootTest(webEnvironment = RANDOM_PORT)
@DirtiesContext(classMode = AFTER_EACH_TEST_METHOD)
class IntegrationTests {
    @Autowired
    private lateinit var restTemplate: TestRestTemplate

    @Test
    fun `Deve criar novo produto`() {
        val produto = Produto(null, "Laptop", BigDecimal("2400.00"))

        val resposta = restTemplate.postForEntity<Produto>("/produtos", produto)

        assertThat(resposta.statusCode).isEqualTo(CREATED)
        assertThat(resposta.body)
            .hasId(3)
            .hasNome("Laptop")
            .hasPreco(BigDecimal("2400.00"))
    }

    @Test
    fun `Deve buscar produtos`() {
        val resposta = restTemplate.getForEntity<Array<Produto>>("/produtos")

        assertThat(resposta.statusCode).isEqualTo(OK)
        assertThat(resposta.body?.size).isEqualTo(2)
        assertThat(resposta.body!![0])
            .hasId(1)
            .hasNome("TV")
            .hasPreco(BigDecimal("1600.00"))
        assertThat(resposta.body!![1])
            .hasId(2)
            .hasNome("Display")
            .hasPreco(BigDecimal("800.00"))
    }

    @Test
    fun `Deve atualizar produto`() {
        val produto = Produto(1, "Laptop", BigDecimal("2400.00"))

        val resposta = restTemplate.putForEntity<Produto>("/produtos/1", produto)

        assertThat(resposta.statusCode).isEqualTo(OK)
        assertThat(resposta.body)
            .hasId(1)
            .hasNome("Laptop")
            .hasPreco(BigDecimal("2400.00"))
    }

    @Test
    fun `Deve deletar produto`() {
        val resposta = restTemplate.deleteForEntity<Produto>("/produtos/1")

        assertThat(resposta.statusCode).isEqualTo(OK)
        assertThat(resposta.body)
            .hasId(1)
            .hasNome("TV")
            .hasPreco(BigDecimal("1600.00"))
    }
}